# <span style="color:yellow">**House Price Prediction**</span>

<span style="color:green">**Objective:**</span> To predict the house price using linear regression

### <span style="color:lightblue">**Import Packages**</span>

In [1]:
import pandas as pd
import numpy as np
import plotly.graph_objs as go
import plotly.express as px
import warnings
warnings.filterwarnings("ignore")
import matplotlib.pyplot as plt
%matplotlib inline

### <span style="color:lightblue">**Read Data**</span>

In [2]:
#load the Boston House Pricing dataset 
from sklearn.datasets import load_boston
raw_data = load_boston()
raw_data

{'data': array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
         4.9800e+00],
        [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
         9.1400e+00],
        [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
         4.0300e+00],
        ...,
        [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
         5.6400e+00],
        [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
         6.4800e+00],
        [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
         7.8800e+00]]),
 'target': array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9, 15. ,
        18.9, 21.7, 20.4, 18.2, 19.9, 23.1, 17.5, 20.2, 18.2, 13.6, 19.6,
        15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 21. , 12.7, 14.5, 13.2,
        13.1, 13.5, 18.9, 20. , 21. , 24.7, 30.8, 34.9, 26.6, 25.3, 24.7,
        21.2, 19.3, 20. , 16.6, 14.4, 19.4, 19.7, 20.5, 25. , 23.4, 18.9,
        35.4, 24.7, 3

In [3]:
raw_data.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename', 'data_module'])

### <span style="color:lightblue">**Understanding Data**</span>

In [4]:
print(raw_data.DESCR)

.. _boston_dataset:

Boston house prices dataset
---------------------------

**Data Set Characteristics:**  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pu

In [5]:
print(raw_data.feature_names)

['CRIM' 'ZN' 'INDUS' 'CHAS' 'NOX' 'RM' 'AGE' 'DIS' 'RAD' 'TAX' 'PTRATIO'
 'B' 'LSTAT']


In [6]:
print(raw_data.data)

[[6.3200e-03 1.8000e+01 2.3100e+00 ... 1.5300e+01 3.9690e+02 4.9800e+00]
 [2.7310e-02 0.0000e+00 7.0700e+00 ... 1.7800e+01 3.9690e+02 9.1400e+00]
 [2.7290e-02 0.0000e+00 7.0700e+00 ... 1.7800e+01 3.9283e+02 4.0300e+00]
 ...
 [6.0760e-02 0.0000e+00 1.1930e+01 ... 2.1000e+01 3.9690e+02 5.6400e+00]
 [1.0959e-01 0.0000e+00 1.1930e+01 ... 2.1000e+01 3.9345e+02 6.4800e+00]
 [4.7410e-02 0.0000e+00 1.1930e+01 ... 2.1000e+01 3.9690e+02 7.8800e+00]]


In [7]:
print(raw_data.target)

[24.  21.6 34.7 33.4 36.2 28.7 22.9 27.1 16.5 18.9 15.  18.9 21.7 20.4
 18.2 19.9 23.1 17.5 20.2 18.2 13.6 19.6 15.2 14.5 15.6 13.9 16.6 14.8
 18.4 21.  12.7 14.5 13.2 13.1 13.5 18.9 20.  21.  24.7 30.8 34.9 26.6
 25.3 24.7 21.2 19.3 20.  16.6 14.4 19.4 19.7 20.5 25.  23.4 18.9 35.4
 24.7 31.6 23.3 19.6 18.7 16.  22.2 25.  33.  23.5 19.4 22.  17.4 20.9
 24.2 21.7 22.8 23.4 24.1 21.4 20.  20.8 21.2 20.3 28.  23.9 24.8 22.9
 23.9 26.6 22.5 22.2 23.6 28.7 22.6 22.  22.9 25.  20.6 28.4 21.4 38.7
 43.8 33.2 27.5 26.5 18.6 19.3 20.1 19.5 19.5 20.4 19.8 19.4 21.7 22.8
 18.8 18.7 18.5 18.3 21.2 19.2 20.4 19.3 22.  20.3 20.5 17.3 18.8 21.4
 15.7 16.2 18.  14.3 19.2 19.6 23.  18.4 15.6 18.1 17.4 17.1 13.3 17.8
 14.  14.4 13.4 15.6 11.8 13.8 15.6 14.6 17.8 15.4 21.5 19.6 15.3 19.4
 17.  15.6 13.1 41.3 24.3 23.3 27.  50.  50.  50.  22.7 25.  50.  23.8
 23.8 22.3 17.4 19.1 23.1 23.6 22.6 29.4 23.2 24.6 29.9 37.2 39.8 36.2
 37.9 32.5 26.4 29.6 50.  32.  29.8 34.9 37.  30.5 36.4 31.1 29.1 50.
 33.3 3

### <span style="color:lightblue">**Preparing the Dataset**</span>

In [8]:
# Creating a dataframe 
dataset = pd.DataFrame(raw_data.data,columns=raw_data.feature_names)
dataset

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4    0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786  1.0  273.0   
502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875  1.0  273.0   
503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675  1.0  273.0   
504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889  1.0  273.0   
505  0.04741   0.0  11.93   0.0  0.573  6.030  80.8  2.5050  1.0  273.0   

     PTRATIO       B  LSTAT  
0       15.3  396.90   4.98  
1       17.8  396.90   9.14  
2       17.8  392.83   4.03  
3       18.7  394.63   2.94  
4       18.7  396.90   5.33  
..       ...     ...    ...  
501     21.0  391.99   9.67  
502     21.0  396.90   9.08  
503     21.0  396.90   5.64  
504     21.0  393.45   6.48  
505     21.0  396.90   7.88  

[506 rows x 13 columns]

In [9]:
dataset['Price'] = raw_data.target

In [10]:
dataset.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  Price  
0     15.3  396.90   4.98   24.0  
1     17.8  396.90   9.14   21.6  
2     17.8  392.83   4.03   34.7  
3     18.7  394.63   2.94   33.4  
4     18.7  396.90   5.33   36.2

In [11]:
dataset.describe()

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032   
std     28.148861    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT       Price  
count  506.000000  506.000000  
mean    12.653063   22.532806  
std      7.141062    9.197104  
min      1.730000    5.000000  
25%      6.950000   17.025000  
50%     11.360000   21.200000  
75%     16.955000   25.000000  
max     37.970000   50.000000

In [12]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    float64
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  Price    506 non-null    float64
dtypes: float64(14)
memory usage: 55.5 KB


In [13]:
dataset['CHAS'].value_counts()

0.0    471
1.0     35
Name: CHAS, dtype: int64

In [14]:
# Checking if there are any missing values
dataset.isnull().sum()

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
Price      0
dtype: int64

### <span style="color:lightblue">**Exploratory Data Analysis**</span>

In [15]:
#Correlation between independent and dependent variables
dataset.corr()

CRIM        ZN     INDUS      CHAS       NOX        RM       AGE  \
CRIM     1.000000 -0.200469  0.406583 -0.055892  0.420972 -0.219247  0.352734   
ZN      -0.200469  1.000000 -0.533828 -0.042697 -0.516604  0.311991 -0.569537   
INDUS    0.406583 -0.533828  1.000000  0.062938  0.763651 -0.391676  0.644779   
CHAS    -0.055892 -0.042697  0.062938  1.000000  0.091203  0.091251  0.086518   
NOX      0.420972 -0.516604  0.763651  0.091203  1.000000 -0.302188  0.731470   
RM      -0.219247  0.311991 -0.391676  0.091251 -0.302188  1.000000 -0.240265   
AGE      0.352734 -0.569537  0.644779  0.086518  0.731470 -0.240265  1.000000   
DIS     -0.379670  0.664408 -0.708027 -0.099176 -0.769230  0.205246 -0.747881   
RAD      0.625505 -0.311948  0.595129 -0.007368  0.611441 -0.209847  0.456022   
TAX      0.582764 -0.314563  0.720760 -0.035587  0.668023 -0.292048  0.506456   
PTRATIO  0.289946 -0.391679  0.383248 -0.121515  0.188933 -0.355501  0.261515   
B       -0.385064  0.175520 -0.356977  0.048788 -0.380051  0.128069 -0.273534   
LSTAT    0.455621 -0.412995  0.603800 -0.053929  0.590879 -0.613808  0.602339   
Price   -0.388305  0.360445 -0.483725  0.175260 -0.427321  0.695360 -0.376955   

              DIS       RAD       TAX   PTRATIO         B     LSTAT     Price  
CRIM    -0.379670  0.625505  0.582764  0.289946 -0.385064  0.455621 -0.388305  
ZN       0.664408 -0.311948 -0.314563 -0.391679  0.175520 -0.412995  0.360445  
INDUS   -0.708027  0.595129  0.720760  0.383248 -0.356977  0.603800 -0.483725  
CHAS    -0.099176 -0.007368 -0.035587 -0.121515  0.048788 -0.053929  0.175260  
NOX     -0.769230  0.611441  0.668023  0.188933 -0.380051  0.590879 -0.427321  
RM       0.205246 -0.209847 -0.292048 -0.355501  0.128069 -0.613808  0.695360  
AGE     -0.747881  0.456022  0.506456  0.261515 -0.273534  0.602339 -0.376955  
DIS      1.000000 -0.494588 -0.534432 -0.232471  0.291512 -0.496996  0.249929  
RAD     -0.494588  1.000000  0.910228  0.464741 -0.444413  0.488676 -0.381626  
TAX     -0.534432  0.910228  1.000000  0.460853 -0.441808  0.543993 -0.468536  
PTRATIO -0.232471  0.464741  0.460853  1.000000 -0.177383  0.374044 -0.507787  
B        0.291512 -0.444413 -0.441808 -0.177383  1.000000 -0.366087  0.333461  
LSTAT   -0.496996  0.488676  0.543993  0.374044 -0.366087  1.000000 -0.737663  
Price    0.249929 -0.381626 -0.468536 -0.507787  0.333461 -0.737663  1.000000

#### <span style="color:red">**DATA VISUALIZATION**</span>

<span style="color:yellow">**Scatter Plot for Crime Rate**</span>

In [16]:
fig = px.scatter(dataset,x='CRIM',y='Price',title="Crime Rate Vs House Price")
fig.show()

we can see that as the crime in a area increases the house price decreases, which means that the crime and house price are inversely correlated with each other and have a negative correlation of -0.388305	

<span style="color:yellow">**Scatter Plot for proportion of residential land zoned for lots over 25,000 sq.ft.**</span>


In [17]:
fig = px.scatter(dataset,x='ZN',y='Price',title="ZN vs Price")
fig.show()

<span style="color:yellow">**Scatter Plot for proportion of non-retail business acres per town**</span>

In [18]:
fig = px.scatter(dataset,x='INDUS',y='Price',title="Indus Vs Price")
fig.show()

<span style="color:yellow">**Scatter Plot for Charles River dummy variable**</span>

Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
       

In [19]:
fig = px.scatter(dataset,x='CHAS',y='Price',title="CHAS Vs Price")
fig.show()

<span style="color:yellow">**Scatter Plot for nitric oxides concentration**</span>
 
nitric oxides concentration (parts per 10 million)

In [20]:
fig = px.scatter(dataset,x='NOX',y='Price',title="NOX Vs Price")
fig.show()

<span style="color:yellow">**Scatter Plot for Rooms per dwelling**</span>

In [21]:
fig = px.scatter(dataset,x='RM',y='Price',title="Rooms per dwelling Vs Price")
fig.show()

So here as we can see if there are more rooms the price will increase

<span style="color:yellow">**Scatter Plot for proportion of owner-occupied units built prior to 1940**</span>

In [22]:
fig = px.scatter(dataset,x='AGE',y='Price',title="AGE  Vs Price")
fig.show()

<span style="color:yellow">**Scatter Plot for Distance to five Boston Employment centeres**</span>


In [23]:
fig = px.scatter(dataset,x='DIS',y='Price',title="DIS  Vs Price")
fig.show()

<span style="color:yellow">**Scatter plot for index of accessibility to radial highways**<span>


In [24]:
fig = px.scatter(dataset,x='RAD',y='Price',title="Radial Highway Access Vs Price")
fig.show()

<span style="color:yellow">**Scatter plot for full-value property-tax rate per $10,000**<span>

In [25]:
fig = px.scatter(dataset,x='TAX',y='Price',title="TAX Vs Price")
fig.show()

<span style="color:yellow">**Scatter plot for pupil-teacher ratio by town**<span>

In [26]:
fig = px.scatter(dataset,x='PTRATIO',y='Price',title="Pupil_teacher_ratio Vs Price")
fig.show()

<span style="color:yellow">**Scatter Plot of Percent lower status of the population**<span>

In [27]:
fig = px.scatter(dataset,x='LSTAT',y='Price',title="LSTAT Vs Price")
fig.show()

In [28]:
dataset

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4    0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786  1.0  273.0   
502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875  1.0  273.0   
503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675  1.0  273.0   
504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889  1.0  273.0   
505  0.04741   0.0  11.93   0.0  0.573  6.030  80.8  2.5050  1.0  273.0   

     PTRATIO       B  LSTAT  Price  
0       15.3  396.90   4.98   24.0  
1       17.8  396.90   9.14   21.6  
2       17.8  392.83   4.03   34.7  
3       18.7  394.63   2.94   33.4  
4       18.7  396.90   5.33   36.2  
..       ...     ...    ...    ...  
501     21.0  391.99   9.67   22.4  
502     21.0  396.90   9.08   20.6  
503     21.0  396.90   5.64   23.9  
504     21.0  393.45   6.48   22.0  
505     21.0  396.90   7.88   11.9  

[506 rows x 14 columns]

In [29]:
# creating independent dataset
independent_dataset = dataset.iloc[:,0:-1]
independent_dataset

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4    0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786  1.0  273.0   
502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875  1.0  273.0   
503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675  1.0  273.0   
504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889  1.0  273.0   
505  0.04741   0.0  11.93   0.0  0.573  6.030  80.8  2.5050  1.0  273.0   

     PTRATIO       B  LSTAT  
0       15.3  396.90   4.98  
1       17.8  396.90   9.14  
2       17.8  392.83   4.03  
3       18.7  394.63   2.94  
4       18.7  396.90   5.33  
..       ...     ...    ...  
501     21.0  391.99   9.67  
502     21.0  396.90   9.08  
503     21.0  396.90   5.64  
504     21.0  393.45   6.48  
505     21.0  396.90   7.88  

[506 rows x 13 columns]

In [30]:
# creating dependent dataset
dependent_dataset = pd.DataFrame()
dependent_dataset['Price'] = dataset.iloc[:,-1]
dependent_dataset

Price
0     24.0
1     21.6
2     34.7
3     33.4
4     36.2
..     ...
501   22.4
502   20.6
503   23.9
504   22.0
505   11.9

[506 rows x 1 columns]

### <span style="color:lightblue">**Train and Test Split**</span>

In [31]:
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test = train_test_split(independent_dataset,dependent_dataset,test_size=0.3,random_state=42)


In [32]:
X_train

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS   RAD    TAX  \
5     0.02985   0.0   2.18   0.0  0.458  6.430  58.7  6.0622   3.0  222.0   
116   0.13158   0.0  10.01   0.0  0.547  6.176  72.5  2.7301   6.0  432.0   
45    0.17142   0.0   6.91   0.0  0.448  5.682  33.8  5.1004   3.0  233.0   
16    1.05393   0.0   8.14   0.0  0.538  5.935  29.3  4.4986   4.0  307.0   
468  15.57570   0.0  18.10   0.0  0.580  5.926  71.0  2.9084  24.0  666.0   
..        ...   ...    ...   ...    ...    ...   ...     ...   ...    ...   
106   0.17120   0.0   8.56   0.0  0.520  5.836  91.9  2.2110   5.0  384.0   
270   0.29916  20.0   6.96   0.0  0.464  5.856  42.1  4.4290   3.0  223.0   
348   0.01501  80.0   2.01   0.0  0.435  6.635  29.7  8.3440   4.0  280.0   
435  11.16040   0.0  18.10   0.0  0.740  6.629  94.6  2.1247  24.0  666.0   
102   0.22876   0.0   8.56   0.0  0.520  6.405  85.4  2.7147   5.0  384.0   

     PTRATIO       B  LSTAT  
5       18.7  394.12   5.21  
116     17.8  393.30  12.04  
45      17.9  396.90  10.21  
16      21.0  386.85   6.58  
468     20.2  368.74  18.13  
..       ...     ...    ...  
106     20.9  395.67  18.66  
270     18.6  388.65  13.00  
348     17.0  390.94   5.99  
435     20.2  109.85  23.27  
102     20.9   70.80  10.63  

[354 rows x 13 columns]

In [33]:
y_train

Price
5     28.7
116   21.2
45    19.3
16    23.1
468   19.1
..     ...
106   19.5
270   21.1
348   24.5
435   13.4
102   18.6

[354 rows x 1 columns]

In [34]:
X_test

CRIM    ZN  INDUS  CHAS    NOX     RM    AGE     DIS   RAD    TAX  \
173  0.09178   0.0   4.05   0.0  0.510  6.416   84.1  2.6463   5.0  296.0   
274  0.05644  40.0   6.41   1.0  0.447  6.758   32.9  4.0776   4.0  254.0   
491  0.10574   0.0  27.74   0.0  0.609  5.983   98.8  1.8681   4.0  711.0   
72   0.09164   0.0  10.81   0.0  0.413  6.065    7.8  5.2873   4.0  305.0   
452  5.09017   0.0  18.10   0.0  0.713  6.297   91.8  2.3682  24.0  666.0   
..       ...   ...    ...   ...    ...    ...    ...     ...   ...    ...   
441  9.72418   0.0  18.10   0.0  0.740  6.406   97.2  2.0651  24.0  666.0   
23   0.98843   0.0   8.14   0.0  0.538  5.813  100.0  4.0952   4.0  307.0   
225  0.52693   0.0   6.20   0.0  0.504  8.725   83.0  2.8944   8.0  307.0   
433  5.58107   0.0  18.10   0.0  0.713  6.436   87.9  2.3158  24.0  666.0   
447  9.92485   0.0  18.10   0.0  0.740  6.251   96.6  2.1980  24.0  666.0   

     PTRATIO       B  LSTAT  
173     16.6  395.50   9.04  
274     17.6  396.90   3.53  
491     20.1  390.11  18.07  
72      19.2  390.91   5.52  
452     20.2  385.09  17.27  
..       ...     ...    ...  
441     20.2  385.96  19.52  
23      21.0  394.54  19.88  
225     17.4  382.00   4.63  
433     20.2  100.19  16.22  
447     20.2  388.52  16.44  

[152 rows x 13 columns]

In [35]:
y_test

Price
173   23.6
274   32.4
491   13.6
72    22.8
452   16.1
..     ...
441   17.1
23    14.5
225   50.0
433   14.3
447   12.6

[152 rows x 1 columns]

In [36]:
original_X_test = X_test

### <span style="color:lightblue">**Standardize Dataset**</span>

since all the features are measured in different units there scale is not same , so while we are using gradient descent approch to find the global minimum if are data is not scaled we should scale it for better convergance to the global minimum.

In [37]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

In [38]:
X_train = scaler.fit_transform(X_train)
X_train

array([[-0.41425879, -0.50512499, -1.29214218, ...,  0.18727079,
         0.39651419, -1.01531611],
       [-0.40200818, -0.50512499, -0.16208345, ..., -0.21208981,
         0.3870674 , -0.05366252],
       [-0.39721053, -0.50512499, -0.60948856, ..., -0.16771641,
         0.42854113, -0.31132373],
       ...,
       [-0.41604586,  3.03838247, -1.3166773 , ..., -0.56707702,
         0.35987906, -0.90549329],
       [ 0.92611293, -0.50512499,  1.00549958, ...,  0.8528718 ,
        -2.87841346,  1.52750437],
       [-0.39030549, -0.50512499, -0.37135358, ...,  1.16348561,
        -3.32828832, -0.25218837]])

In [39]:
# Apply transoformation on test data
X_test = scaler.transform(X_test)
X_test

array([[-0.406801  , -0.50512499, -1.02225586, ..., -0.74457062,
         0.41241246, -0.47605794],
       [-0.41105674,  1.26662874, -0.68165068, ..., -0.30083661,
         0.42854113, -1.25185755],
       [-0.4051199 , -0.50512499,  2.39678516, ...,  0.8084984 ,
         0.35031705,  0.79535229],
       ...,
       [-0.35439903, -0.50512499, -0.71195877, ..., -0.38958342,
         0.25688594, -1.09697922],
       [ 0.25423453, -0.50512499,  1.00549958, ...,  0.8528718 ,
        -2.98970133,  0.53487511],
       [ 0.77732457, -0.50512499,  1.00549958, ...,  0.8528718 ,
         0.33199949,  0.56585078]])

### <span style="color:lightblue">**Model Training**</span>

In [40]:
from sklearn.linear_model import LinearRegression
reg_inst = LinearRegression()
reg_inst.fit(X_train,y_train)

LinearRegression()

In [41]:
print(reg_inst.intercept_)
print(reg_inst.coef_)

[23.01581921]
[[-1.10834602  0.80843998  0.34313466  0.81386426 -1.79804295  2.913858
  -0.29893918 -2.94251148  2.09419303 -1.44706731 -2.05232232  1.02375187
  -3.88579002]]


### <span style="color:lightblue">**Prediction**</span>

In [42]:
y_test['Predicted_house_price'] = reg_inst.predict(X_test)

In [43]:
y_test

Price  Predicted_house_price
173   23.6              28.648960
274   32.4              36.495014
491   13.6              15.411193
72    22.8              25.403213
452   16.1              18.855280
..     ...                    ...
441   17.1              17.403672
23    14.5              13.385941
225   50.0              39.983425
433   14.3              16.682863
447   12.6              18.285618

[152 rows x 2 columns]

In [44]:
test_data = pd.concat([original_X_test,y_test[['Price','Predicted_house_price']]],axis=1)
test_data

CRIM    ZN  INDUS  CHAS    NOX     RM    AGE     DIS   RAD    TAX  \
173  0.09178   0.0   4.05   0.0  0.510  6.416   84.1  2.6463   5.0  296.0   
274  0.05644  40.0   6.41   1.0  0.447  6.758   32.9  4.0776   4.0  254.0   
491  0.10574   0.0  27.74   0.0  0.609  5.983   98.8  1.8681   4.0  711.0   
72   0.09164   0.0  10.81   0.0  0.413  6.065    7.8  5.2873   4.0  305.0   
452  5.09017   0.0  18.10   0.0  0.713  6.297   91.8  2.3682  24.0  666.0   
..       ...   ...    ...   ...    ...    ...    ...     ...   ...    ...   
441  9.72418   0.0  18.10   0.0  0.740  6.406   97.2  2.0651  24.0  666.0   
23   0.98843   0.0   8.14   0.0  0.538  5.813  100.0  4.0952   4.0  307.0   
225  0.52693   0.0   6.20   0.0  0.504  8.725   83.0  2.8944   8.0  307.0   
433  5.58107   0.0  18.10   0.0  0.713  6.436   87.9  2.3158  24.0  666.0   
447  9.92485   0.0  18.10   0.0  0.740  6.251   96.6  2.1980  24.0  666.0   

     PTRATIO       B  LSTAT  Price  Predicted_house_price  
173     16.6  395.50   9.04   23.6              28.648960  
274     17.6  396.90   3.53   32.4              36.495014  
491     20.1  390.11  18.07   13.6              15.411193  
72      19.2  390.91   5.52   22.8              25.403213  
452     20.2  385.09  17.27   16.1              18.855280  
..       ...     ...    ...    ...                    ...  
441     20.2  385.96  19.52   17.1              17.403672  
23      21.0  394.54  19.88   14.5              13.385941  
225     17.4  382.00   4.63   50.0              39.983425  
433     20.2  100.19  16.22   14.3              16.682863  
447     20.2  388.52  16.44   12.6              18.285618  

[152 rows x 15 columns]

In [45]:
residuals = y_test['Price'] - y_test['Predicted_house_price']
residuals

173    -5.048960
274    -4.095014
491    -1.811193
72     -2.603213
452    -2.755280
         ...    
441    -0.303672
23      1.114059
225    10.016575
433    -2.382863
447    -5.685618
Length: 152, dtype: float64

### <span style="color:lightblue">**Visualize Predicted and Actual Values**</span>

In [46]:
fig = px.scatter(test_data,x='Predicted_house_price',y='Price',color_discrete_sequence=['red'])
fig.show()

Since the predictions are linear we can say our model performed well

### <span style="color:lightblue">**Model Evaluation**</span>

In [47]:
from sklearn.metrics import mean_squared_error

MSE = mean_squared_error(y_test['Price'],y_test['Predicted_house_price'])
print("Mean squared error: %.2f"%MSE)


Mean squared error: 21.52


In [48]:
from sklearn.metrics import mean_absolute_error
MAE = mean_absolute_error(y_test['Price'],y_test['Predicted_house_price'])
print("Mean Absolute error: %.2f"%MAE)

Mean Absolute error: 3.16


In [49]:
RSME = np.sqrt(MSE)
print("Root Mean squared error: %.2f"%RSME)

Root Mean squared error: 4.64


In [50]:
from sklearn.metrics import r2_score
R_square = r2_score(y_test['Price'],y_test['Predicted_house_price'])
print("R squared error: %.2f"%R_square)

R squared error: 0.71


### New Data Prediction

In [51]:
single_row_data = dataset.iloc[0:1,0:-1]
single_row_data

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE   DIS  RAD    TAX  PTRATIO  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.09  1.0  296.0     15.3   

       B  LSTAT  
0  396.9   4.98

In [52]:
single_row_data.shape

(1, 13)

In [53]:
#transformation of new data 
single_row_data = scaler.transform(single_row_data)
single_row_data

array([[-0.41709233,  0.29216419, -1.27338003, -0.28154625, -0.16513629,
         0.34715902, -0.13030059,  0.15267164, -0.97798895, -0.66777595,
        -1.32142483,  0.42854113, -1.04769976]])

In [54]:
predicted_value = reg_inst.predict(single_row_data)
predicted_value

array([[30.08649576]])

### Pickel The Model File

In [55]:
import pickle

In [56]:
pickle.dump(reg_inst,open('house_price_regmodel.pkl','wb'))

In [57]:
# Loading the pickle file for testing
pickel_model = pickle.load(open('house_price_regmodel.pkl','rb'))

In [58]:
# predict with loaded model 
test_value = pickel_model.predict(single_row_data)
test_value

array([[30.08649576]])